### ISSUE TITLE
Comments can affect the formatting behaviour in unintentional ways

### ISSUE THREAD

**Issue by garyb** [natefaubion/purescript-tidy#16]
``` purescript
module Main where

changeDialog = do
  -- oh noes
  { x } ← H.get
  pure unit
```

The presence of the comment reformats the record destructuring like this:

``` purescript
module Main where

changeDialog = do
  -- oh noes
  { x
  } ← H.get
  pure unit
```

**Comment by natefaubion** [comment]
This is a problem with delimiters and leading comments in general. The formatter doesn't look at spans, but it does bubble up a "multiline" flag. Essentially, printing the leading token makes it mark it as "multiline" because of the leading comment.

**Comment by natefaubion** [comment]
To be more accurate, this crops up when a semantic element gets `flexGroup`ed for dynamic layout behavior, but has leading line comments. At the point it gets grouped, the comments have already been baked into the `FormatDoc`, and so it opts out of dynamic layout behavior since it already knows it's broken up over multiple lines. The problem is we want the `flexGroup` to apply to the semantic element, and not the comment elements. The "correct" solution is probably to buffer comments in `FormatDoc`, and only bake them in when joining two `FormatDoc`s. There's potentially a quick fix though to just reassociate this grouping for the common delimiters (`formatList`). There may be other cases that fit this bill, but I'm pretty confident that 90% of the time this might crop up would be with the primary list/delimiter formatter.

For example, applications are fine:
```
wat =
  -- ok
  a b c
```

**Comment by natefaubion** [comment]
I've put in the quick fix in https://github.com/natefaubion/purescript-tidy/commit/400d99e4fc799d792e58fd57f82aa45b8ebb30cf. This should fix the issue with delimiters for now, but there may be other cases that require the general fix. I'll leave this open in case other issues like it pop up.

**Comment by natefaubion** [comment]
Another _case_ of this is comments on branches in a multi-case.
```purescript
case _, _ of
  -- A comment
  foo, bar -> ...
```
Will get turned into
```purescript
case _, _ of
  -- A comment
  foo,
  bar -> ...
```

**Comment by andys8** [comment]
I'm seeing a similar case.

```purescript
-- ids
data NodeId'
type NodeId = Id NodeId'

data LinkId'
type LinkId = Id LinkId'
```

With the `--ids` comment, purs-tidy will split the first two lines and separate them with a newwline. This won't happen for the second id type because there is no comment

```purescript
-- ids
data NodeId'

type NodeId = Id NodeId'

data LinkId'
type LinkId = Id LinkId'
```

**Comment by natefaubion** [comment]
@andys8 Thanks! That's a good example, and yeah, it's caused by the same root issue. I think both that and the binder issue can be corrected without rewriting FormatDoc. I'll probably go down that route as long as possible 😭.

**Comment by natefaubion** [comment]
The proper fix may be a lot easier now though since I rewrote the `Hang` module.

**Comment by natefaubion** [comment]
Fixed in 0.5.3
